File: python/P515.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    # bfs
    def largestValues(self, root):
        if not root:
            return []
        curr_list = [root]
        answer = []
        while curr_list:
            maxnumber = None
            next_level = []
            for each in curr_list:
                maxnumber = max(each.val, maxnumber)
                if each.left:
                    next_level.append(each.left)
                if each.right:
                    next_level.append(each.right)
            answer.append(maxnumber)
            curr_list = next_level
        return answer
        

File: java/P515.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    // dfs
    public List<Integer> largestValues(TreeNode root) {
        List<Integer> answer = new ArrayList<Integer>();
        dfs(root, answer, 0);
        return answer;
    }
    
    
    private void dfs(TreeNode root, List<Integer> answer, int d) {
        if (root == null) {
            return;
        }
        if (d == answer.size()) {
            answer.add(root.val);
        }
        else {
            answer.set(d, Math.max(root.val, answer.get(d)));
        }
        dfs(root.left, answer, d+1);
        dfs(root.right, answer, d+1);
    }
}